test/Generics/inverse_casting_availability.swift

```diff
@@ -0,0 +1,117 @@
+// RUN: %target-typecheck-verify-swift \
+// RUN:   -debug-diagnostic-names -target arm64-apple-macos14.4 \
+// RUN:   -enable-experimental-feature NonescapableTypes
+
+protocol P {}
+struct NCG<T: ~Copyable> {}
+extension NCG: P where T: Copyable {} // expected-note 2{{requirement_implied_by_conditional_conformance}}
+
+struct NEG<T: ~Escapable> {}
+extension NEG: P {} // expected-note {{requirement_implied_by_conditional_conformance}}
+
+struct All {}
+struct NoCopy: ~Copyable {}
+struct NoEscape: ~Escapable {}
+
+
+
+/// MARK: dynamic casts are gated by availability. Older runtimes don't check
+/// for conformance to Copyable so they'll give bogus results.
+
+// expected-note@+1 8{{availability_add_attribute}}
+func dyn_cast_errors<T: ~Copyable, V: ~Escapable>(
+                    _ generic: NCG<T>, _ concrete: NCG<NoCopy>,
+                    _ genericEsc: NEG<V>, _ concreteEsc: NEG<NoEscape>) {
+  _ = concrete as? any P // expected-error {{availability_copyable_generics_casting_only_version_newer}} // expected-note {{availability_guard_with_version_check}}
+  _ = generic as? any P // expected-error {{availability_copyable_generics_casting_only_version_newer}} // expected-note {{availability_guard_with_version_check}}
+
+  _ = concrete is any P // expected-error {{availability_copyable_generics_casting_only_version_newer}} // expected-note {{availability_guard_with_version_check}}
+  _ = generic is any P // expected-error {{availability_copyable_generics_casting_only_version_newer}} // expected-note {{availability_guard_with_version_check}}
+
+  _ = concrete as! any P // expected-error {{availability_copyable_generics_casting_only_version_newer}} // expected-note {{availability_guard_with_version_check}}
+  _ = generic as! any P // expected-error {{availability_copyable_generics_casting_only_version_newer}} // expected-note {{availability_guard_with_version_check}}
+
+  _ = genericEsc as? any P // expected-error {{availability_escapable_generics_casting_only_version_newer}} // expected-note {{availability_guard_with_version_check}}
+  _ = concreteEsc is any P // expected-error {{availability_escapable_generics_casting_only_version_newer}} // expected-note {{availability_guard_with_version_check}}
+}
+
+@available(macOS 15, *)
```

Comment:
The test will need to only run on Apple platforms in order to test availability diagnostics. I suggest using `@available(SwiftStdlib 6.0, *)` here to cover all the platforms, and then restrict the test like this:

```
// REQUIRES: OS=macosx || OS=ios || OS=tvos || OS=watchOS || OS=xros
```

Reply:
thank you!